Add JetNews to the available workloads

luis-machado-arm opened the pull request:
These three commits add the JetNews workload to the main branch of wa. The first two commits were contributed by Jakub Szostakowski. The third commit is where I picked things up to start making it useful for our team's needs. And also to keep these workloads in a central repo/branch so they don't get scattered around multiple teams/repos.

luis-machado-arm commented on wa/workloads/jetnews/__init__.py:
```diff
@@ -0,0 +1,48 @@
+# limitations under the License.
+#
+
+from wa import Workload
```
Thanks for the review. Given the fact I haven't actually implemented this code (someone else did), I'd say it might be unfamiliarity with how WA works and its modules. I'll have to touch these up and make them do the right thing. So I might have a few questions here and there, as I get familiar with this code.

luis-machado-arm commented on wa/workloads/jetnews/__init__.py:
```diff
@@ -0,0 +1,48 @@
+        super(JetNews, self).run(context)
+
+        # Run the first set of basic tests:
+        tests = ["ScrollArticleTest",
```
Definitely. That's what I have in mind as we expand on those tests, to have some parameters to select types of tests and maybe loop count. This is an initial version, basically what the original code/author did.

luis-machado-arm commented on wa/workloads/jetnews/wa_jetnews_base.yaml:
```diff
@@ -0,0 +1,18 @@
+config:
```
Got it. Will fix that up. And will likely update the docs as well.

luis-machado-arm commented on wa/workloads/jetnews/__init__.py:
```diff
@@ -0,0 +1,48 @@
+    def teardown(self, context):
+        super(JetNews, self).teardown(context)
+        for package in self.package_names:
+            self.target.execute(f"am force-stop {package}")
```
Nice! That would be great. Thanks for the info.

luis-machado-arm commented on wa/workloads/jetnews/__init__.py:
```diff
@@ -0,0 +1,48 @@
+            self.target.execute(
+                "am instrument -w "
+                "-e class com.arm.test." + test + " "
+                "com.arm.benchmark/androidx.test.runner.AndroidJUnitRunner"
```
Great. I'll give these changes a try.

douglas-raillard-arm commented on wa/workloads/jetnews/__init__.py:
```diff
@@ -0,0 +1,48 @@
+    package_names = ["com.example.jetnews", "com.arm.benchmark"]
+
+    def setup(self, context):
+        super(JetNews, self).setup(context)
```
There is no need to pass any parameter to `super()` in this context, we are in Python 3

luis-machado-arm replied:
Thanks. I made this change now.

douglas-raillard-arm commented on wa/workloads/jetnews/__init__.py:
```diff
@@ -0,0 +1,48 @@
+    def setup(self, context):
+        super(JetNews, self).setup(context)
+        for package in self.package_names:
+            self.target.execute(f"am force-stop {package}")
```
nit: use `shlex.quote()` to ensure the value of `package` is not expanded by the shell. This won't matter given the current values, but that is required in general.

luis-machado-arm replied:
I'll have that in mind. With Marc's suggestion, this code is gone, since it uses ApkWorkload.

douglas-raillard-arm commented on wa/workloads/jetnews/__init__.py:
```diff
@@ -0,0 +1,48 @@
+        # Invoke each test one after the other.
+        for test in tests:
+            self.target.execute(
+                "am instrument -w "
```
Maybe WA does things differently but I'd recommend using a single line here, otherwise that breaks the grep-ability of the code.

luis-machado-arm replied:
I switched to putting the command in a variable (on a single-line) and then using the variable to invoke the command.

douglas-raillard-arm commented on wa/workloads/jetnews/__init__.py:
```diff
@@ -0,0 +1,48 @@
+        for test in tests:
+            self.target.execute(
+                "am instrument -w "
+                "-e class com.arm.test." + test + " "
```
can use f-strings instead of concatenation

luis-machado-arm replied:
Done now. Thanks!

luis-machado-arm commented:
Addressed the comments in this updated PR. I plan to submit the documentation bits separately. Honestly, I don't think the workloads is yet at a stage where we can just use it. It needs a bit more work.

marcbonnici commented:
> Addressed the comments in this updated PR

Thanks for fixing these, it looks like the changes are heading in the right direction.

> So I might have a few questions here and there, as I get familiar with this code.

No problem please reach out if you have any questions and I can try to answer or point you to the relevant parts of the [documentation](https://workload-automation.readthedocs.io/en/latest/developer_information.html#) that might help.

> Honestly, I don't think the workloads is yet at a stage where we can just use it. It needs a bit more work.

Makes sense, on my side I was able to execute the workload and it launched the Jetnews application, however it doesn't currently seem to perform any tests. There are also no test metrics collected from the workload that would allow for comparison between runs. Are these the sort of extra work that you have planned for this workload?

luis-machado-arm commented:
> > Honestly, I don't think the workloads is yet at a stage where we can just use it. It needs a bit more work.
> 
> Makes sense, on my side I was able to execute the workload and it launched the Jetnews application, however it doesn't currently seem to perform any tests. There are also no test metrics collected from the workload that would allow for comparison between runs. Are these the sort of extra work that you have planned for this workload?

Yes, exactly. This initial contribution is a base from which we should (hopefully) build something more meaningful. So things like improved/more meaningful tests, collecting metrics etc, will come later. I don't think we want dependencies on perfetto, so I still need to check how that is going to work in terms of data collection. We'll probably do ftrace, but we need the jank information.

Right now there are 3 very basic tests that scroll things, and you can select those and pass the number of iterations. Even those tests don't seem to work 100% at the moment.

marcbonnici commented:
Ok cool thanks for confirming. In which case happy to leave this here as a base that can be used for further development.

luis-machado-arm commented:
I have a different patch to enable this.
